server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java

```diff
@@ -37,38 +39,70 @@
     private static final String DEVICE_2_ID = MockKeyProvider.DEVICE_2;
 
-    @InjectMocks
     private final JwtBearerAssertionTokenAuthenticator tokenAuthenticator = new JwtBearerAssertionTokenAuthenticator(
             AUDIENCE, 15);
 
-    @Mock
-    private ClientDetailsService clientDetailsService;
+    private ClientDetailsService clientDetailsService = Mockito.mock(ClientDetailsService.class);
     
     private long currentTimeSecs;
 
     @Before
     public void beforeMethod() {
-        MockitoAnnotations.initMocks(this);
         this.tokenAuthenticator.setClientPublicKeyProvider(new MockKeyProvider());
-        when(this.clientDetailsService.loadClientByClientId(DEVICE_1_ID))
-                .thenReturn(new BaseClientDetails(DEVICE_1_ID, null, null, null, null, null));
+        BaseClientDetails testUaaClient = new BaseClientDetails(DEVICE_1_CLIENT_ID, null, null, null, null, null);
+        testUaaClient.addAdditionalInformation(ClientConstants.ALLOWED_DEVICE_ID, DEVICE_1_ID);
+        when(this.clientDetailsService.loadClientByClientId(DEVICE_1_CLIENT_ID)) .thenReturn(testUaaClient);
         this.currentTimeSecs =  System.currentTimeMillis()/1000;
     }
 
     @Test
     public void testSuccess() {
         long currentTime = System.currentTimeMillis();
-        String token = new MockAssertionToken().mockAssertionToken(DEVICE_1_ID, currentTime- 240000L,
-                600, TENANT_ID, AUDIENCE);
+        String token = new MockAssertionToken().mockAssertionToken(DEVICE_1_CLIENT_ID, DEVICE_1_ID,
+                currentTime, 600, TENANT_ID, AUDIENCE);
         String header = new MockClientAssertionHeader().mockSignedHeader(this.currentTimeSecs, DEVICE_1_ID, TENANT_ID);
         this.tokenAuthenticator.setClientDetailsService(this.clientDetailsService);
         Authentication authn = this.tokenAuthenticator.authenticate(token, header, MockKeyProvider.DEVICE1_PUBLIC_KEY);
-        Assert.assertEquals(DEVICE_1_ID, authn.getPrincipal());
+        Assert.assertEquals(DEVICE_1_CLIENT_ID, authn.getPrincipal());
         Assert.assertEquals(true, authn.isAuthenticated());
     }
+    
+    // c1.allowed_device_id is not set
+    @Test(expected=AuthenticationException.class)
+    public void testNoAllowedDeviceIdSet() {
+        long currentTime = System.currentTimeMillis();
+        String token = new MockAssertionToken().mockAssertionToken(DEVICE_1_CLIENT_ID, DEVICE_1_ID,
+                currentTime, 600, TENANT_ID, AUDIENCE);
+        String header = new MockClientAssertionHeader().mockSignedHeader(this.currentTimeSecs, DEVICE_1_ID, TENANT_ID);
+        
+        //allowed_device_id not set
+        when(this.clientDetailsService.loadClientByClientId(DEVICE_1_CLIENT_ID)) 
+            .thenReturn(new BaseClientDetails(DEVICE_1_CLIENT_ID, null, null, null, null, null));
+        this.tokenAuthenticator.setClientDetailsService(this.clientDetailsService);
+        
```

Comment:
I agree that we need to have json validation even before we proceed with authenticating jwt assertion.  During validation we should check that header and assertion contain all the expected claims and claims are not null and not empty.  
In this case we are validating uaa client attribute.  We test if expected attribute is not set and if it's set but not authorized for the device.  We should also test it expected attribute is set to null or empty.